TypeScript (#1819)
enkidevs/curriculum: typescript/introduction/tsc-data-types/tsc-intro-data-types.md

---
author: kapnobatai136

type: normal

category: must-know

aspects:
  - introduction

---

# Data Types

---
## Content

> A data type is defined by the values it can take and the operations that can be performed on it and . For example, `number` is a data type.

Reviewer: ````suggestion
A data type is defined by the values it can take and the operations that can be performed on it. For example, a `number` is a data type that can be subtracted but a `string` is a sequence of characters and it cannot be subtracted.

```ts
let num = 5;
let ok = num - 3; // no error

let str = 'enki';
let nono = str - 3; 
//         ^^^ 
// error because we're subtracting a string
```
````

> In TypeScript, as the name already suggests, data types are a big deal. Let's take a look at the `hello.ts` file we wrote in the first workout:

Reviewer: ```suggestion
We can define types for any value (think variable) in TypeScript code. Let's take a look at the `hello.ts` file we wrote in the first workout:
```

```ts
// hello.ts
const hello = (name: string) => {
  console.log(`Hello ${name}!`)
}

> let myName = 'Andrei';
> hello(myName);

Reviewer: ```suggestion
hello('Andrei');
```

```

> Notice that when writing the arguments of the `hello` function, we wrote `name: string`. This means that you can only pass `strings` as arguments, and no other type. If we were to change `myName` to a different type:

Reviewer: Parameters are definitions of values in a function signature, arguments are values passed as parameters.

```suggestion
Notice that the `name` parameter in the `hello` function has a type definition `string` next to it. This means that you can only pass a `strings` as the value for `name`, and no other type. 

If we were to pass a different type we would get an error:
```

> ```ts
> let myName = 120;
> ```
> 
> When compiling the code, TypeScript will throw the following error:
> 
> ```
> error TS2345: Argument of type 'number' is not assignable to parameter of type 'string'.
> ```

Reviewer: ````suggestion
```ts
hello(123);
// error TS2345: Argument of type 'number' is not assignable to parameter of type 'string'.
```
````

In the following insights we will go over the possible data types and how to use them.

---
## Practice

> Which file do you think will compile?

Reviewer: ```suggestion
Which file do you think will compile without errors?
```
